File: 03 Record Power System/Gather.py
from PyP100 import PyP100
import threading
import json
from datetime import datetime

duration = int(60 * 60 * 8.5) # In Seconds
filename = "data/power.json"

smartPlugIP = "0.0.0.0" # Add Smart Plug IP Address Here
smartPlugEmail = "email@email.com" # Add email used for Smart Plug account here
smartPlugPassword = "password" # Add password used for Smart Plug account here

p100 = PyP100.P100(smartPlugIP, smartPlugEmail, smartPlugPassword)
p100.handshake()
p100.login()

startTime = datetime.now().strftime("%H:%M:%S")
data = {"startTime": startTime, "errors": 0, "power": [], "time": []}
errors = 0
isRecTime = "true"

def getCurrentPower():
    try:
        info = p100.getEnergy()
        print(info["result"]["current_power"], datetime.now().strftime("%H:%M:%S.%f"), int(((duration * 2) - len(data["power"]))/2))
        data["power"].append(info["result"]["current_power"])
        if(isRecTime == "true"):
            data["time"].append(datetime.now().strftime("%H:%M:%S.%f"))
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=4)
            
    except Exception as e:
        global errors
        errors += 1
        data["errors"] += 1
        if(len(data["power"]) == 0):
            data["power"].append(0)
            if(isRecTime == "true"):
                data["time"].append(datetime.now().strftime("%H:%M:%S.%f"))
        else:
            data["power"].append(data["power"][len(data["power"]) - 1])
            if(isRecTime == "true"):
                data["time"].append(datetime.now().strftime("%H:%M:%S.%f"))
        print("ERROR " + str(errors) + ": " + str(e))

def getDataLoop():
    if(len(data["power"]) < ((duration * 2))):
        threading.Timer(0.5, getDataLoop).start()
        getCurrentPower()


print(duration)
getDataLoop()

File: 03 Record Power System/Gather.js
import fs from "fs";
import { loginDeviceByIp, getEnergyUsage } from "tp-link-tapo-connect";

const smartPlugEmail = "email@email.com"; // Add email used for Smart Plug account here
const smartPlugPassword = "password"; // Add password used for Smart Plug account here
const smartPlugIP = "0.0.0.0"; // Add Smart Plug IP Address Here

const duration = Math.floor(8.5 * 60 * 60 * 1000); // In Milliseconds
const filename = "data/power.json";

const startTime = getCurrentTime("h:m:s");
var data = { startTime: startTime, power: [], time: [] };

console.log("Duration: ", duration / 1000 + "s");

async function run() {
	const device_token = await loginDeviceByIp(smartPlugEmail, smartPlugPassword, smartPlugIP);
	if (!device_token?.deviceIp) return;

	const interval = setInterval(async () => {
		const { current_power } = await getEnergyUsage(device_token);
		console.log(current_power);
		data.power.push(current_power);
		data.time.push(getCurrentTime("h:m:s:ms"));
	}, 500 - 8);

	setTimeout(() => {
		clearInterval(interval);
		fs.writeFileSync(filename, JSON.stringify(data), "utf8");
	}, duration + 100);
}
run();

function getCurrentTime(format) {
	return format
		.split(":")
		.map((formatItem) => {
			switch (formatItem) {
				case "h":
					const hours = new Date().getHours();
					return hours < 10 ? "0" + hours : hours;
					break;
				case "m":
					const minutes = new Date().getMinutes();
					return minutes < 10 ? "0" + minutes : minutes;
					break;
				case "s":
					const seconds = new Date().getSeconds();
					return seconds < 10 ? "0" + seconds : seconds;
					break;
				case "ms":
					const milliseconds = new Date().getMilliseconds();
					return milliseconds < 100 ? (milliseconds < 10 ? "00" + milliseconds : "0" + milliseconds) : milliseconds;
					break;
			}
		})
		.join(":");
}
